### ISSUE TITLE
In Riverpod 1.0.0-dev.7 ProviderObserver.didUpdateProvider newValue and previousValue always have same value when using StateProvider

### ISSUE THREAD

**Issue by rydmike** [rrousselGit/river_pod#693]
## Description 

When using Riverpod 1.0.0-dev.7 changing a `StateProvider`'s state now triggers the `ProviderObserver.didUpdateProvider` correctly (it did not in previous versions). However, the `newValue` and `previousValue` are always equal, the `previousValue` never contains the previous state or value.

Relate issue: https://github.com/rrousselGit/river_pod/issues/694

## To Reproduce

Please see this simple counter example repo: https://github.com/rydmike/pods_obs_issue

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(
    ProviderScope(
      observers: <ProviderObserver>[ProvidersObserver()],
      child: MyApp(),
    ),
  );
}

final StateProvider<int> counterProvider =
    StateProvider<int>((StateProviderRef ref) {
  return 0;
}, name: 'counter');

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Riverpod 1.0.0-dev.7 Issue'),
    );
  }
}

class MyHomePage extends ConsumerWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '${ref.watch(counterProvider).state}',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          ref.read(counterProvider).state++;
        },
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

/// ProvidersObserver used in this demo to debugPrint changes.
class ProvidersObserver extends ProviderObserver {
  @override
  void didUpdateProvider(ProviderBase<dynamic> provider, Object? previousValue,
      Object? newValue, ProviderContainer container) {
    if (newValue is StateController && previousValue is StateController) {
      debugPrint('Provider is StateController: '
          '${provider.name ?? provider.runtimeType} \n'
          'new value: ${newValue.state}\n'
          'previous value: ${previousValue.state}');
    }
  }
}
```

When clicking the (+) button the observer is called and we see console debug prints.

```
Restarted application in 536ms.
flutter: Provider is StateController: counter 
flutter: new value: 1
flutter: previous value: 1
flutter: Provider is StateController: counter 
flutter: new value: 2
flutter: previous value: 2
flutter: Provider is StateController: counter 
flutter: new value: 3
flutter: previous value: 3
```

## Expected behavior

When clicking the (+) button we expect the previous value to contain the actual previous value.

```
Restarted application in 591ms.
flutter: Provider is StateController: counter 
flutter: new value: 1
flutter: previous value: 0
flutter: Provider is StateController: counter 
flutter: new value: 2
flutter: previous value: 1
flutter: Provider is StateController: counter 
flutter: new value: 3
flutter: previous value: 2
```

**Comment by rrousselGit** [comment]
This is expected.

The function is receiving the StateController as parameter, which never changes. It is mutated.

**Comment by rydmike** [comment]
Sure, but from an observer usability point of view it makes no sense with a StateProvider.

**Comment by rrousselGit** [comment]
This is no different to how the other providers works.
Say you're using ChangeNotifierProvider, the previous and new value will be the same too.

**Comment by rydmike** [comment]
In combination with https://github.com/rrousselGit/river_pod/issues/694 this will force us to maintain the previous state separately so we can avoid doing things in the observer, if the state of the StateProvider was not changed. Not very usable considering the intended purpose of the observer.

**Comment by rrousselGit** [comment]
> Not very usable considering the intended purpose of the observer.

What intended purpose?

As I said, this is the expected behaviour and is nothing new. This is how it always worked.

**Comment by rydmike** [comment]
Hi @TimWhiting thanks for the reply. 

## TLDR
Yes I know and I of course I can do that. I was hoping to avoid that, since it requires a lot of rewrites, but sure totally doable. 

## Details

For the case in question there are probably +100 independent individual values (bools, ints, double, enums, strings) that should be observed and logged individually when its value changes, and only when its value is changed and only if it actually changed, regardless of how and where it is changed in the app, using any available method it can be changed with.

Additionally all these parameter values all have related UI widgets to control them, and when each one is modified it should only rebuild itself and trigger rebuilds of things that depends on its individual value and nothing else, and observe the change, and also if its value is changed anywhere elsewhere in code, same thing. 

Using individual StateProviders for these parameters seemed like a handy and simple way of doing this. So I was testing the approach in an experiment. I like to try and see how to me unfamiliar things works and scale (not just performance wise, but from a dev experience viewpoint too) in practice before using it in any real solution.

I know, I can easily add an extension like you suggested. If I do so, I would still have to update probably 200 places where it should be used. Ok not a big thing either, doable sure. However, I can not enforce the extension usage. So at that point, I might as well make classes for all the values and extend StateNotifer and use StateNotiferProvider and do whatever I like/need with them. At this point I would also consider if and how many of them would/should be bundled in same class and use filters in Widgets to minimize the earlier described rebuilds, to only the property value in question (with a StateProivder for each value I got that automatically). For this use case and experiment, I was trying to avoid the code overhead of all this. 

In this particular case, all functionality each of these Dart built-in type based, property values needed, was to be able to rebuild only widgets dependent on each value in question, observe and log changes to the value made anywhere in the app, if it was changed somehow with any means it can be changed with. Plus only do this if the value was actually changed, setting it to same value as before is not considered a change.

Thus using StateProviders and the observer seemed like a nice reasonable fit, having with minimal extra dev overhead fuzz. I found that yes, it actually works very well, and was super quick and easy to setup and use. Except for the part that setting a StateProvider's state to a new value, that is same as the value it had before, also triggers the observer, even if nothing changed so there was not really anything new to observe and hence log either.

## Observer docs

I tried this simplified approach as an experiment, partly because the docs say this:

https://riverpod.dev/docs/concepts/provider_observer
![image](https://user-images.githubusercontent.com/39990307/130875735-bd84ac99-77a6-4142-89b1-afa3fa7eb87b.png)

and API doc comment:

```dart
/// An object that listens to the changes of a [ProviderContainer].
///
/// This can be used for logging or making devtools.
abstract class ProviderObserver {
```
I wrongly though it would be fine for logging the changes in the StateProviders, and only logging them when changes in their values actually occur.

The way the Observer works now does not make it entirely suitable for the mentioned logging in the docs, since it gets called also when there has been no change in its value, that you would typically like to log, but only when the value actually changed. 

In the experiment and later intended usage, the observer is just used to observe changes in many (100+) StateProvider changes, it actually works very well. It is unusual that they get set to same value they had before, so it is thus **very** rare that it logs a change, that was not actually a change, and when it happens, it does not really matter so much if the same value is logged an extra time anyway. That is why I said the solution actually works and does what I hoped it would do, still it is good to know that the observer is not really suitable if you want to truly log only unique value changes.

## Conclusion

Anyway, thanks both of you for the replies and good suggestions. In certain scenarios the experimental setup I described is still very usable as is, but it is also good to know now when it is not.

Knowing this, I can certainly solve it with a totally different approach that works for more stringent logging requirements too (the Riverpod observer is not usable for that, in its current form at least). Any such solution requires a bit more code, but yeah its just more code... 🤷🏻‍♂️  😃 👍🏻

**Comment by rrousselGit** [comment]
I think you're complaining about the wrong thing. The issue isn't ProviderObserver.
What you dislike is StateProvider. You're looking for a way to not update dependents when the new and previous values are ==.

If that's what you want, you could subclass StateController to override the `set state`:

```dart
class EqualController<T> extends StateController<T> {
  EqualController(T value): super(value);

  @override
  set state(T newState) {
    if (newState != state) super.state = newState;
  }
}
```

**Comment by rydmike** [comment]
OK sure 🙂👍 
Not entirely sure why it does not do that otb. Come to think of it, I seem to recall another issue asking about that too, or perhaps it was another package 🤷🏼‍♂️ 

Thx as always for great discussions and advise, closing this issue 💙
